### PR TITLE
[FLINK-31346][runtime] IO scheduler does not throw TimeoutException if numRequestedBuffers is greater than 0

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*We currently rely on throw exception to trigger downstream task failover to avoid read buffer request deadlock. But if `numRequestedBuffers` is greater than 0, `IO scheduler` does not throw `TimeoutException`. This will cause a deadlock.*

## Brief change log

  - *Let `IO Schedule` handle un-expected exceptions.*
  - *Fix the bug that `IO scheduler` does not throw `TimeoutException` if `numRequestedBuffers` is greater than 0.*

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by reswqa** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java [hunk 1]
```diff
@@ -436,7 +442,17 @@ private void mayTriggerReading() {
                 && numRequestedBuffers + bufferPool.getNumBuffersPerRequest() <= maxRequestedBuffers
                 && numRequestedBuffers < bufferPool.getAverageBuffersPerRequester()) {
             isRunning = true;
-            ioExecutor.execute(this);
+            ioExecutor.execute(
+                    () -> {
+                        try {
+                            run();
+                        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java]: Good catch.

**Comment by reswqa** [comment]
Thanks @TanYuxin-tyx for the review, I have updated this according to your comments, please take a look again.

**Comment by TanYuxin-tyx** [comment]
Now LGTM, thanks @reswqa for fixing the issue.

**Review by TanYuxin-tyx — APPROVED** [review]

**Pull request closed without merging**